devices/Bq2579x/Bq2579x.cs

```diff
@@ -0,0 +1,939 @@
+        /// Gets battery present status.
+        /// </summary>
+        /// <value><see langword="true"/> if VBAT present, otherwise <see langword="false"/>.</value>
+        [Property]
+        public bool VbatPresent => GetChargerStatus2().HasFlag(ChargerStatus2.VbatPresent);
+
+        /// <summary>
+        /// Gets Input Current Optimizer (ICO) status.
+        /// </summary>
+        [Property]
+        public IcoStatus IcoStatus => GetIcoStatus();
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="Bq2579x" /> class.
+        /// </summary>
+        /// <param name="i2cDevice"><see cref="I2cDevice"/> to communicate with Si7021 device.</param>
+        /// <exception cref="InvalidOperationException">When failing to read part information.</exception>
+        /// <exception cref="NotSupportedException">If the part information returned is invalid, thus the connected part is not one of the supported BQ2579x devices.</exception>
+        /// <exception cref="ArgumentNullException">If <paramref name="i2cDevice"/> is null.</exception>
+        public Bq2579x(I2cDevice i2cDevice)
+        {
+            _i2cDevice = i2cDevice ?? throw new ArgumentNullException();
+
+            // read part information
+            ReadPartInformation();
+        }
+
+        /// <summary>
+        /// Read part information to perform a sanity check for the correct part.
+        /// </summary>
+        private void ReadPartInformation()
+        {
+            byte[] buffer = ReadFromRegister(Register.REG48_Part_Information, 1);
+
+            _deviceModel = (Model)(buffer[0] & DevicePartNumberMask);
+
+            // sanity check for valid part
+            if (_deviceModel != Model.Bq25792
+                && _deviceModel != Model.Bq25798)
+            {
+                throw new NotSupportedException();
+            }
+        }
+
+        /// <summary>
+        /// Reset I2C device watchdog.
+        /// </summary>
+        /// <exception cref="InvalidOperationException">If the I2C operation to write to the device to reset the watchdog fails.</exception>
+        public void ResetWatchdog()
+        {
+            WriteToRegister(Register.REG10_Charger_Control_1, 0b0000_1000);
+        }
+
+        /// <inheritdoc/>
+        public void Dispose()
+        {
+            _i2cDevice?.Dispose();
+            _i2cDevice = null;
+        }
+
+        #region REG1B_Charger_Status_0
+
+        // REG1B_Charger_Status_0 Register
+        // | 7 6 5 4 3 2 1 0 |
+        // | IINDPM_STAT | VINDPM_STAT | WD_STAT | RESERVED | PG_STAT | AC2_PRESENT_STAT | AC1_PRESENT_STAT | VBUS_PRESENT_STAT |
+        ////
+
+        private ChargerStatus0 GetChargerStatus0()
+        {
+            byte[] buffer = ReadFromRegister(Register.REG1C_Charger_Status_1, 1);
+
+            return (ChargerStatus0)buffer[0];
+        }
+
+        #endregion
+
+        #region REG1C_Charger_Status_1
+
+        // REG1C_Charger_Status_1 Register
```

Comment:
Thanks a lot for adding all those in the comments, it makes the life easier when it comes to understanding what the code does and this will be super useful for maintenance.